storefront-bff/src/main/java/com/yas/storefrontbff/filter/CartAuthenticationFilter.java

```diff
@@ -0,0 +1,105 @@
+    private final ServiceUrlConfig serviceUrlConfig;
+    private final ReactiveClientRegistrationRepository clientRegistrationRepository;
+    private static final String GRANT_TYPE = "grant_type";
+
+    public CartAuthenticationFilter(WebClient webClient,
+                                    ServiceUrlConfig serviceUrlConfig,
+                                    ReactiveClientRegistrationRepository clientRegistrationRepository) {
+        this.webClient = webClient;
+        this.serviceUrlConfig = serviceUrlConfig;
+        this.clientRegistrationRepository = clientRegistrationRepository;
+    }
+
+    @Override
+    @NonNull
+    public Mono<Void> filter(ServerWebExchange exchange, WebFilterChain chain) {
+
+        ServerHttpRequest request = exchange.getRequest();
+        if (request.getPath().toString().contains("/api/cart/storefront/carts")
+                && Objects.equals(request.getMethod(), HttpMethod.POST)) {
+            AtomicBoolean isAuthenticated = new AtomicBoolean(false);
+
+            return ReactiveSecurityContextHolder.getContext()
+                    .map(SecurityContext::getAuthentication)
+                    .map(Authentication::getPrincipal)
+                    .doOnNext(principal -> isAuthenticated.set(true))
+                    .doOnError(Throwable::printStackTrace)
+                    .then(Mono.defer(() -> {
+                        if (!isAuthenticated.get()) {
+                            MultiValueMap<String, String> formData = new LinkedMultiValueMap<>();
+                            ClientRegistration clientRegistration = clientRegistrationRepository.findByRegistrationId("keycloak").block();
+                            formData.add(GRANT_TYPE, "client_credentials");
+                            formData.add("client_id", clientRegistration.getClientId());
+                            formData.add("client_secret", clientRegistration.getClientSecret());
+
+                            return getToken(formData).map(clientToken -> createGuestUser(clientToken.access_token())
+                                    .map(createdUserVm -> {
+                                        formData.remove(GRANT_TYPE);
+                                        formData.add(GRANT_TYPE, "password");
+                                        formData.add("username", createdUserVm.email());
+                                        formData.add("password", createdUserVm.password());
+                                        return getToken(formData).doOnNext(userToken -> exchange
+                                                .getRequest()
+                                                .mutate()
+                                                .headers(h -> h.setBearerAuth(userToken.access_token()))
+                                                .build()).thenReturn(exchange);
+                                    })).flatMap(monox3Void -> monox3Void.flatMap(monox2Void -> monox2Void.flatMap(chain::filter)));
```

Comment:
Can we auto authenticate the user by set value to SecurityContextHolder, so that subsequence request don't need to do this again